src/main/java/org/opensearch/security/configuration/SecurityConfigVersionHandler.java

```diff
@@ -0,0 +1,330 @@
+import org.opensearch.security.securityconf.DynamicConfigFactory;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.security.securityconf.impl.SecurityDynamicConfiguration;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.client.Client;
+
+import org.greenrobot.eventbus.Subscribe;
+
+import static org.opensearch.security.support.ConfigConstants.SECURITY_CONFIGURATIONS_VERSIONS_ENABLED;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_CONFIGURATIONS_VERSIONS_ENABLED_DEFAULT;
+
+/**
+ * Manages security configuration versioning in OpenSearch.
+ */
+
+public class SecurityConfigVersionHandler {
+
+    private final int maxVersionsToKeep;
+
+    private static final Logger log = LogManager.getLogger(SecurityConfigVersionHandler.class);
+    private final Client client;
+    private final String SecurityConfigVersionsIndex;
+    private final SecurityConfigVersionsLoader configVersionsLoader;
+
+    private final ConfigurationRepository cr;
+    private final Settings settings;
+    private final ThreadContext threadContext;
+    private final ThreadPool threadPool;
+
+    public SecurityConfigVersionHandler(
+        ConfigurationRepository cr,
+        Settings settings,
+        ThreadContext threadContext,
+        ThreadPool threadPool,
+        Client client
+    ) {
+        this.cr = cr;
+        this.settings = settings;
+        this.threadContext = threadContext;
+        this.client = client;
+        this.SecurityConfigVersionsIndex = settings.get(
+            ConfigConstants.SECURITY_CONFIG_VERSIONS_INDEX_NAME,
+            ConfigConstants.OPENDISTRO_SECURITY_DEFAULT_CONFIG_VERSIONS_INDEX
+        );
+        this.configVersionsLoader = new SecurityConfigVersionsLoader(client, settings);
+        this.threadPool = threadPool;
+        this.maxVersionsToKeep = settings.getAsInt(
+            ConfigConstants.SECURITY_CONFIG_VERSION_RETENTION_COUNT,
+            ConfigConstants.SECURITY_CONFIG_VERSION_RETENTION_COUNT_DEFAULT
+        );
+    }
+
+    @Subscribe
+    public void onConfigInitialized(DynamicConfigFactory.SecurityConfigChangeEvent event) {
+        if (!isVersionIndexEnabled(settings)) return;
+
+        try {
+            log.info("Initializing version index ({})", SecurityConfigVersionsIndex);
+
+            if (!createOpendistroSecurityConfigVersionsIndexIfAbsent()) {
+                log.info("Version index already exists, skipping initialization.");
+            }
+
+            waitForOpendistroSecurityConfigVersionsIndexToBeAtLeastYellow();
+
+            String nextVersionId = fetchNextVersionId();
+            User user = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+            String userinfo = (user != null) ? user.getName() : ("v1".equals(nextVersionId) ? "system" : "unknown");
```

Comment:
why distinguish between "system" and "unknown" here?

If the user is empty here it either means its 1) the initial seed of the index on node bootstrap, 2) securityadmin tool is being used or 3) user is making calls with admin certificate (2 + 3 are the same)

In the case of using the securityadmin tool, will that create multiple entries in the version history or 1 for the whole script invocation?